SDL/PixelTriplet.cu

```diff
@@ -433,78 +442,128 @@ __device__ bool SDL::passPT3RZChiSquaredCuts(struct modules& modulesInGPU, uint1
+__device__ float SDL::computePT3RZChiSquared(struct modules& modulesInGPU, uint16_t* lowerModuleIndices, float* rtPix, float* xPix, float* yPix, float* zPix, float* rts, float* xs, float* ys, float* zs, float pixelSegmentPt, float pixelSegmentPx, float pixelSegmentPy, float pixelSegmentPz, int pixelSegmentCharge, float pixelG, float pixelF, float pixelRadiusPCA)
 { 
     float slope = (zPix[1] - zPix[0])/(rtPix[1] - rtPix[0]);
     float residual = 0;
     float error = 0;
     //hardcoded array indices!!!
     float RMSE = 0;
+
+    float Pt=pixelSegmentPt, Px=pixelSegmentPx, Py=pixelSegmentPy, Pz=pixelSegmentPz;
+    int charge=pixelSegmentCharge;
+    float CenterX=pixelG/100, CenterY=pixelF/100, CircleRadius=pixelRadiusPCA/100; // for calculating initial state corrections
+    float x0 = xPix[0]/100,x1 = xPix[1]/100;
+    float y0 = yPix[0]/100,y1 = yPix[1]/100;
+    float z0 = zPix[0]/100,z1 = zPix[1]/100;
+    float r0 = rtPix[0]/100,r1 = rtPix[1]/100;
+
+    float B = 3.8112;
+    float a = -0.299792*B*charge;
+
     for(size_t i = 0; i < 3; i++)
     {
+        float xsi = xs[i]/100;
+        float ysi = ys[i]/100;
+        float zsi = zs[i]/100;
+        float rtsi = rts[i]/100;
         uint16_t lowerModuleIndex = lowerModuleIndices[i];
         const int moduleType = modulesInGPU.moduleType[lowerModuleIndex];
         const int moduleSide = modulesInGPU.sides[lowerModuleIndex];
         const int moduleSubdet = modulesInGPU.subdets[lowerModuleIndex];
 
-        residual = moduleSubdet == SDL::Barrel ? (zs[i] - zPix[0]) - slope * (rts[i] - rtPix[0]) : (rts[i] - rtPix[0]) - (zs[i] - zPix[0])/slope;
- 
+        float diffr,diffz;
+        float p = sqrt(Px*Px+Py*Py+Pz*Pz);
+
+        float rou = a/p;
+        if (moduleSubdet == SDL::Endcap){
+            float s = (zsi-z1)*p/Pz;
+            float x = x1 + Px/a*sin(rou*s)-Py/a*(1-cos(rou*s));
+            float y = y1 + Py/a*sin(rou*s)+Px/a*(1-cos(rou*s));
+//          float z = z1+Pz/p*s;
+            diffr = fabs(rtsi-sqrt(x*x+y*y))*100;
+        }
+
+        if (moduleSubdet == SDL::Barrel){
+            float paraA = r1*r1 + 2*(Px*Px+Py*Py)/(a*a) + 2*(y1*Px-x1*Py)/a - rtsi*rtsi;
+            float paraB = 2*(x1*Px+y1*Py)/a;
+            float paraC = 2*(y1*Px-x1*Py)/a+2*(Px*Px+Py*Py)/(a*a);
+            //termA+paraB*sin(\rho s)  = paraC*sqrt(1-sin(\rho s)*sin(\rho s))
+            float A=paraB*paraB+paraC*paraC;
+            float B=2*paraA*paraB;
+            float C=paraA*paraA-paraC*paraC;
+            //A*sin(\rho s)*sin(\rho s)+B*sin(\rho s)+C=0;
+            float sol1 = (-B+sqrt(B*B-4*A*C))/(2*A);
+            float sol2 = (-B-sqrt(B*B-4*A*C))/(2*A);
+            float solz1 = asin(sol1)/rou*Pz/p+z1;
+            float solz2 = asin(sol2)/rou*Pz/p+z1;
+            float diffz1 = fabs(solz1-zsi)*100;
+            float diffz2 = fabs(solz2-zsi)*100;
+            diffz = min(diffz1,diffz2);
```

Comment:
Thanks a lot for these calculations, Yanxi. I think it might be a good idea to have a comment in the code pointing to slides 19-21 in your presentation (https://indico.cern.ch/event/1185895/contributions/4982756/attachments/2526561/4345805/helix%20pT3%20summarize.pdf), so that we know how and why these all these computations are there.

Reply:
Thanks for this suggestion! I have added the link~